src/main/java/io/elastic/jdbc/ColumnNamesProvider.java

```diff
@@ -1,113 +1,131 @@
+    result.add("out", outMetadata.build()).add("in", inMetadata.build());
+    return result.build();
+  }
 
-    public JsonObject getMetaModel(JsonObject configuration) {
-        JsonObject result = new JsonObject();
-        JsonObject inMetadata = new JsonObject();
-        JsonObject properties = getColumns(configuration);
-        inMetadata.addProperty("type", "object");
-        inMetadata.add("properties", properties);
-        result.add("out", inMetadata);
-        result.add("in", inMetadata);
-        return result;
+  public JsonObject getColumns(JsonObject configuration) {
+    if (configuration.getString("tableName") == null || configuration.getString("tableName")
+        .isEmpty()) {
+      throw new RuntimeException("Table name is required");
     }
+    String tableName = configuration.getString("tableName");
+    JsonObjectBuilder properties = Json.createObjectBuilder();
+    Connection connection = null;
+    ResultSet rs = null;
+    String schemaName = null;
+    boolean isEmpty = true;
+    Boolean isOracle = (configuration.getString("dbEngine").equals("oracle")) ? true : false;
+    Boolean isMssql = (configuration.getString("dbEngine").equals("mssql")) ? true : false;
+    try {
+      connection = Utils.getConnection(configuration);
+      DatabaseMetaData dbMetaData = connection.getMetaData();
+      rs = dbMetaData.getPrimaryKeys(null, null, tableName);
+      if (tableName.contains(".")) {
+        schemaName = tableName.split("\\.")[0];
+        tableName = tableName.split("\\.")[1];
+      }
+      rs = dbMetaData.getColumns(null, schemaName, tableName, "%");
+      while (rs.next()) {
+        JsonObjectBuilder field = Json.createObjectBuilder();
+        String name = rs.getString("COLUMN_NAME");
+        Boolean isRequired = false;
+        if (isMssql) {
+          String isAutoincrement =
+              (rs.getString("IS_AUTOINCREMENT") != null) ? rs.getString("IS_AUTOINCREMENT") : "";
+          Integer isNullable = (rs.getObject("NULLABLE") != null) ? rs.getInt("NULLABLE") : 1;
+          isRequired = isNullable == 0 && !isAutoincrement.equals("YES");
+        } else {
+          isRequired = false;
+        }
+        field.add("required", isRequired)
+            .add("title", name)
+            .add("type", convertType(rs.getInt("DATA_TYPE")));
+        properties.add(name, field.build());
+        isEmpty = false;
+      }
+      if (isEmpty) {
+        properties.add("empty dataset", "no columns");
+      }
 
-    public JsonObject getColumns(JsonObject configuration) {
-        if (configuration.get("tableName") == null || configuration.get("tableName").getAsString().isEmpty()) {
-            throw new RuntimeException("Table name is required");
+    } catch (SQLException e) {
+      throw new RuntimeException(e);
+    } finally {
+      if (rs != null) {
+        try {
+          rs.close();
+        } catch (SQLException e) {
+          logger.error("Failed to close result set {}", e.toString());
```

Comment:
You don`t need to call toString() from exception

Reply:
done